Puppeteer/Playwright course (fixed branch name) (#357)
apify/apify-docs: content/academy/puppeteer_playwright/reading_intercepting_requests.md

---
title: V - Reading & intercepting requests
description: You can use DevTools, but did you know that you can do all the same stuff (plus more) programmatically? Read and intercept requests in Puppeteer/Playwright.
menuWeight: 7.5
paths:
    - puppeteer-playwright/reading-intercepting-requests
---

# [](#reading-intercepting-requests) Reading & intercepting requests

On any website that serves up images, makes [XMLHttpRequests](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest), or fetches content in some other way, you can see those requests (and their responses) in the [**Network** tab]({{@link api_scraping/general_api_scraping/locating_and_learning.md}}) of your browser's DevTools. Lots of data about the request can be found there, such as the headers, payload, and response body.

> In Playwright and Puppeteer, it is also possible to read (and even intercept) requests being made on the page after its initial load - programmatically. This is very useful for things like reading dynamic headers, saving API responses, blocking certain resources, and much more.

Reviewer: You can read/intercept even the initial request.

```suggestion
In Playwright and Puppeteer, it is also possible to read (and even intercept) requests being made on the page - programmatically. This is very useful for things like reading dynamic headers, saving API responses, blocking certain resources, and much more.
```

import puppeteer from 'puppeteer';

const browser = await puppeteer.launch({ headless: false });
const page = await browser.newPage();

// Our code will go here

await page.goto('https://soundcloud.com/tiesto/following');

await page.waitForTimeout(10000);
await browser.close();
</marked-tab>
```

## [](#reading-requests) Reading requests

We can use the [`page.on()`](https://pptr.dev/#?product=Puppeteer&version=v14.0.0&show=api-event-close) function to listen for the **request** event, passing in a callback function. The first parameter of the passed in callback function is an object representing the request.

Upon visiting Tiësto's following page, we can see in the **Network** tab that a request is made to fetch all of the users which he is following.

![Request to grab Tiësto's following list]({{@asset puppeteer_playwright/images/tiesto-request.webp}})

Let's go ahead and listen for this request in our code:

```marked-tabs
<marked-tab header="Playwright" lang="javascript">
// Listen for all requests
page.on('request', (req) => {
    // If the URL doesn't include our keyword, ignore it
    if (!req.url().includes('followings')) return;

>     console.log('Request was made!')

Reviewer: ```suggestion
    console.log('Request for followers was made!')
```

});
</marked-tab>
<marked-tab header="Puppeteer" lang="javascript">
// Listen for all requests
page.on('request', (req) => {
    // If the URL doesn't include our keyword, ignore it
    if (!req.url().includes('followings')) return;

>     console.log('Request was made!')

Reviewer: ```suggestion
    console.log('Request for followers was made!')
```

});
</marked-tab>
```

> Note that you should always define any request reading/interception code prior to calling the `page.goto()` function.

Cool! So now when we run our code, we'll see this logged to the console:

```text

> Request was made!

Reviewer: ```suggestion
Request for followers was made!
```

import puppeteer from 'puppeteer';

const browser = await puppeteer.launch({ headless: false });
const page = await browser.newPage();

// Listen for all requests
page.on('request', (req) => {
    // If the URL doesn't include our keyword, ignore it
    if (!req.url().includes('followings')) return;

    // Convert the request URL into a URL object
    const url = new URL(req.url());

    // Print the search parameters in object form
    console.log(Object.fromEntries(url.searchParams));
});

await page.goto('https://soundcloud.com/tiesto/following');

await page.waitForTimeout(10000);
await browser.close();
</marked-tab>
```

After running this code, we can see this logged to the console:

```text
{
  client_id: 'llCGDUjKpxUslgO1yEce7Zh95PXE78Bo',
  limit: '12',
  offset: '0',
  linked_partitioning: '1',
  app_version: '1652347025',
  app_locale: 'en'
}
```

## [](#reading-responses) Reading responses

Listening for and reading responses is very similar to reading requests. The only difference is that we need to listen for the **response** event instead of **request**. Additionally, the object passed into the callback function represents the response instead of the request.

> <!-- To access the response object from **Route** in Playwright, we need to wait for it with `await route.request().response()`. Nothing else needs to change. -->

Reviewer: old comment?

This time, instead of grabbing the query parameters of the request URL, let's grab hold of the response body and print it to the console in JSON format:

```marked-tabs
<marked-tab header="Playwright" lang="javascript">
// Notice that the callback function is now async
page.on('response', async (res) => {
    if (!res.request().url().includes('followings')) return;

    // Grab the response body in JSON format
    const json = await res.json();
    console.log(json);
});
</marked-tab>
<marked-tab header="Puppeteer" lang="javascript">
// Notice that the callback function is now async
page.on('response', async (res) => {
    if (!res.request().url().includes('followings')) return;

    // Grab the response body in JSON format
    const json = await res.json();
    console.log(json);
});
</marked-tab>
```

Upon running this code, we'll see the API response logged into the console:

![API response in console]({{@asset puppeteer_playwright/images/api-response-tiesto.webp}})

## [](#intercepting-requests) Intercepting requests

> One of the most popular ways of speeding up website loading in Puppeteer and Playwright is by blocking certain resources from loading. These resources are usually CSS files, images, and other miscellanous resources that aren't super necessary (mainly because the computer doesn't have eyes - it doesn't care how the website looks!).

Reviewer: Im totally missing the part about cache being turned off by turning on interception and how that's dangerous.

For blocking requests, there's [a completely different way of doing that that we have in the SDK](https://github.com/apify/apify-js/blob/55e0cce99dc7bfc5c303bb76885cdaf037ff6cd7/src/puppeteer_utils.js#L187) which does not turn off cache. It would be nice to mention the CDP method here as the correct way for blocking requests when performance is important. Also pls mention that it only works in Chromium based browsers.

});

await page.goto('https://soundcloud.com/tiesto/following');

await page.waitForTimeout(10000);
await browser.close();
</marked-tab>
```

After running this code, we can see this logged to the console:

```text
{
  client_id: 'llCGDUjKpxUslgO1yEce7Zh95PXE78Bo',
  limit: '12',
  offset: '0',
  linked_partitioning: '1',
  app_version: '1652347025',
  app_locale: 'en'
}
```

## [](#reading-responses) Reading responses

Listening for and reading responses is very similar to reading requests. The only difference is that we need to listen for the **response** event instead of **request**. Additionally, the object passed into the callback function represents the response instead of the request.

This time, instead of grabbing the query parameters of the request URL, let's grab hold of the response body and print it to the console in JSON format:

```marked-tabs
<marked-tab header="Playwright" lang="javascript">
// Notice that the callback function is now async
page.on('response', async (res) => {
    if (!res.request().url().includes('followings')) return;

    // Grab the response body in JSON format
    const json = await res.json();
    console.log(json);

> });

Reviewer: The res.json() can throw if the response is not JSON, causing an unhandled promise rejection. Having unhandled errors or rejections in the page handlers is quite a common error of devs, so I think it would make sense to add a try catch and use that opportunity to explain why that's important.

</marked-tab>
<marked-tab header="Puppeteer" lang="javascript">
import puppeteer from 'puppeteer';

const browser = await puppeteer.launch({ headless: false });
const page = await browser.newPage();

const blockedExtensions = ['.png', '.css', '.jpg', '.jpeg', '.pdf', '.svg'];

// Enable request interception (skipping this step will result in an error)
await page.setRequestInterception(true);

// Listen for all requests
page.on('request', async (req) => {
    // If the request ends in a blocked extension, abort the request
    if (blockedExtensions.some((str) => req.url().endsWith(str))) return req.abort();
    // Otherwise, continue
    await req.continue();
});

await page.goto('https://soundcloud.com/tiesto/following');

await page.waitForTimeout(10000);
await browser.close();
</marked-tab>
```

> You can also use `request.resourceType()` to grab the resource type.

Here's what we see when we run this logic:

![SoundCloud with no CSS or image resources loaded]({{@asset puppeteer_playwright/images/ugly-soundcloud.webp}})

This confirms that we've successfully blocked the CSS and image resources from loading.

#### [](#quick-note) Quick note about resource blocking

> Something **very** important to note is that by using request interception, the browser's cache is turned **off**. This means that resources on websites that would normally be cached (and pulled from the cache instead on the next request for those resources) will not be cached, which can have varying negative effects on performance, especially when making multiple requests to the same page.

Reviewer: ```suggestion
Something **very** important to note is that by using request interception, the browser's cache is turned **off**. This means that resources on websites that would normally be cached (and pulled from the cache instead on the next request for those resources) will not be cached, which can have varying negative effects on performance, especially when making many requests to the same domain, which is very common in web scraping.
```
